{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from qiime2 import Artifact\n",
    "from Bio import Entrez, SeqIO\n",
    "from gzip import open as gopen\n",
    "from time import sleep\n",
    "from ete3 import Tree, TextFace, CircleFace, NodeStyle, TreeStyle\n",
    "from os import path\n",
    "import glob\n",
    "\n",
    "!mkdir phylogenies\n",
    "\n",
    "Entrez.email = \"amir@adssc.org\"\n",
    "\n",
    "taxa = Artifact.load('dada/taxonomy.qza').view(pd.DataFrame)\n",
    "core = pd.read_csv('metadata/core_asvs.tsv',sep='\\t')\n",
    "summary = pd.read_csv('metadata/summary.tsv',sep='\\t')\n",
    "\n",
    "!wget https://www.arb-silva.de/fileadmin/silva_databases/current/Exports/SILVA_138.1_SSURef_NR99_tax_silva.fasta.gz\n",
    "!zcat SILVA_138.1_SSURef_NR99_tax_silva.fasta.gz | makeblastdb -title silva -out silva -dbtype nucl\n",
    "refs = SeqIO.to_dict(SeqIO.parse(gopen('SILVA_138.1_SSURef_NR99_tax_silva.fasta.gz','rt'),'fasta'))\n",
    "\n",
    "qza = 'dada/rep-seqs.qza'\n",
    "a = !unzip $qza\n",
    "digest = a[1].split('/')[0].replace('  inflating: ','')\n",
    "inf = digest + '/data/dna-sequences.fasta'\n",
    "ASVs = SeqIO.to_dict(SeqIO.parse(inf,'fasta'))\n",
    "!rm -r $digest\n",
    "\n",
    "groups = {\n",
    "    'rhodo': 'Rhodobacterales',\n",
    "    'flavo': 'Flavobacteriales',\n",
    "    'sapro': 'Saprospirales',\n",
    "    'cyto': 'Cytophagales',\n",
    "    'sbr': 'SBR1031',\n",
    "    'cyan': 'Cyanobacteria',\n",
    "    'myxo': 'Myxococcota',\n",
    "    'planc': 'Planctomycetes',\n",
    "    'rhiz': 'Rhizobiales',\n",
    "    'alt': 'Alteromonadales' \n",
    "}\n",
    "\n",
    "for key, group in groups.items():\n",
    "\n",
    "    rhodo = core.loc[core.Order==group].asv.tolist()\n",
    "    rhodo += [j for j in taxa.index if j[:6] in [i[:6] for i in summary.loc[summary.order==group].feature.tolist()][1:]]\n",
    "    rhodo = set(rhodo)\n",
    "    queries = [ASVs[i] for i in rhodo]\n",
    "    q = 'phylogenies/%s.fasta' % key\n",
    "    SeqIO.write(queries,q,'fasta')\n",
    "    blastout = 'phylogenies/%s.out' % key\n",
    "    \n",
    "    a = !blastn -query $q -out $blastout -outfmt 6 -max_hsps 1 -max_target_seqs 50 -db silva\n",
    "\n",
    "    blast = pd.read_csv(blastout,sep='\\t',header=None)\n",
    "    blast =set([i for i in blast[1]])\n",
    "    alnin = 'phylogenies/%s_with_refs.fasta' % key\n",
    "\n",
    "    with open(alnin,'wt') as hndl:\n",
    "        for i in blast:\n",
    "            hndl.write(refs[i].format('fasta'))\n",
    "        hndl.write(open(q,'r').read())\n",
    "\n",
    "    alnout = 'phylogenies/%s_with_refs.aln.fasta' % key\n",
    "    a = !mafft --maxiterate 1000 --localpair --thread 14 $alnin > $alnout\n",
    "\n",
    "    trimout = 'phylogenies/%s_with_refs.trm.fasta' % key\n",
    "    a = !trimal -in $alnout -out $trimout -gt 0.1\n",
    "    a = !raxmlHPC-PTHREADS -s $trimout -f a -x 123 -N 100 -n $key -m GTRGAMMA -p 456 -T 10 -w /home/amir/Dropbox/DropboxOnBioPC91/SeaGrass/phylogenies\n",
    "\n",
    "\n",
    "    with open('ref_seq_sources/%s_sources' % key,'wt') as hndl:\n",
    "        for n in blast:\n",
    "            handle = Entrez.efetch(db=\"nucleotide\", id=n.split('.')[0], rettype=\"gb\", retmode=\"text\")\n",
    "            sleep(1)\n",
    "            r = SeqIO.read(handle,'gb')\n",
    "            handle.close()\n",
    "            if 'isolation_source' in r.features[0].qualifiers:\n",
    "                source = r.features[0].qualifiers['isolation_source'][0]\n",
    "                hndl.write(\"%s\\t%s\\n\" % (n,source))\n",
    "\n",
    "\n",
    "for f in glob.glob('phylogenies/RAxML_bipartitions.*'):\n",
    "    taxon = f.split('.')[-1]\n",
    "    if path.exists('phylogenies/%s.png' % taxon):\n",
    "        continue\n",
    "    t = Tree(f)\n",
    "\n",
    "    ns = NodeStyle()\n",
    "    ns['size'] = 0\n",
    "\n",
    "    t.set_outgroup(t.get_midpoint_outgroup())\n",
    "    \n",
    "    sources = pd.read_csv('ref_seq_sources/%s_sources' % taxon,index_col=0,header=None,names=['source'])\n",
    "    \n",
    "    for n in t.traverse():\n",
    "        n.set_style(ns)\n",
    "        if n.is_leaf():\n",
    "            if n.name in refs:\n",
    "                color = 'black'\n",
    "                desc =   refs[n.name].description.split(';')[-1]\n",
    "                if desc.startswith(taxon):\n",
    "                    color='blue'\n",
    "                n.add_face(TextFace(n.name + ' ',fgcolor='black'),0,position='branch-right')\n",
    "                n.add_face(TextFace(desc,fgcolor=color),1,position='branch-right')\n",
    "                if n in sources.index:\n",
    "                    n.add_face(TextFace(' '+sources.at[n,'source'],fgcolor='green'),2,position='branch-right')\n",
    "            else:\n",
    "                n.add_face(TextFace(taxon+' '+n.name[:6],fgcolor='red'),0,position='branch-right')\n",
    "        elif n.support > 70:\n",
    "            n.add_face(CircleFace(4,'black'),0,position='float')\n",
    "\n",
    "    ts = TreeStyle()\n",
    "    ts.show_leaf_name = False\n",
    "    ts.scale=2000\n",
    "\n",
    "    t.render('phylogenies/%s.png' % taxon, tree_style=ts,dpi=600,w=1500)  \n",
    "    t.render('phylogenies/%s.pdf' % taxon, tree_style=ts,dpi=600,w=1500) "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
